README.rst
.. image:: https://img.shields.io/github/v/release/Dennis-van-Gils/project-Humidistat
    :target: https://github.com/Dennis-van-Gils/project-Humidistat
    :alt: Latest release
.. image:: https://img.shields.io/badge/code%20style-black-000000.svg
    :target: https://github.com/psf/black
.. image:: https://img.shields.io/badge/License-MIT-purple.svg
    :target: https://github.com/Dennis-van-Gils/project-Humidistat/blob/master/LICENSE.txt

Humidistat
==========
*A Physics of Fluids project.*

This project concerns an Arduino powered controller with a graphical user-interface in Python to regulate the relative humidity inside of a chamber. It does so by controlling two air lines, one with moist air and the other with dry air, using solenoid valves. The temperature, pressure and relative humidity of the chamber and the optional upstream mixing vessel are measured and can be logged over time.

- Github: https://github.com/Dennis-van-Gils/project-Humidistat

.. image:: https://raw.githubusercontent.com/Dennis-van-Gils/project-Humidistat/master/screenshot.png

.. image:: https://raw.githubusercontent.com/Dennis-van-Gils/project-Humidistat/master/docs/control_bands_explained.png

Hardware
========
* Adafruit #3857: Adafruit Feather M4 Express - Featuring ATSAMD51 Cortex M4
* Pimoroni PIM472: BME280 Breakout - Temperature, Pressure, Humidity Sensor
* ...

Instructions
============
Download the `latest release <https://github.com/Dennis-van-Gils/project-Humidistat/releases/latest>`_
and unpack to a folder onto your drive.

Flashing the firmware
---------------------

Double click the reset button of the Feather while plugged into your PC. This
will mount a drive called `FEATHERBOOT`. Copy
`src_mcu/_build_Feather_M4/CURRENT.UF2 <https://github.com/Dennis-van-Gils/project-Humidistat/raw/main/src_mcu/_build_Feather_M4/CURRENT.UF2>`_
onto the Featherboot drive. It will restart automatically with the new firmware.

Running the application
-----------------------


Prerequisites
~~~~~~~~~~~~~

| Preferred distribution: Anaconda full or Miniconda

    * `Anaconda <https://www.anaconda.com>`_
    * `Miniconda <https://docs.conda.io/en/latest/miniconda.html>`_

Open `Anaconda Prompt` and navigate to the unpacked folder. Run the following to
install the necessary packages:

::

   cd src_python
   conda update -n base -c defaults conda
   conda create -n humi -c conda-forge  --force -y python=3.8.10
   conda activate humi
   pip install -r requirements.txt

Now you can run the graphical user interface of the humidistat.
In Anaconda prompt:

::

   conda activate humi
   ipython main.py


LED status lights
=================

The RGB LED of the Feather M4 will indicate its status:

* Blue : We're setting up
* Green: All okay and idling

Every read out, the LED will flash brightly turquoise.


CHANGELOG.rst
Changelog
=========

1.2.0 (2023-06-26)
------------------
* Added support for Adafruit Feather M0 Express board

1.1.0 (2022-07-28)
------------------
* Increased significant numbers in temperature and humidity from 1 to 2 digits
* Added time and date information in the header

1.0.0 (2022-02-01)
------------------
* Initial release


src_mcu/src/main.cpp
/*******************************************************************************
  Humidistat v1

  Hardware & pin out
  ------------------

  Adafruit Feather M4 Express
    Sensors:
    - BME280 #1, I2C address 0x76
    - BME280 #2, I2C address 0x77
    Actuators:
    - solenoid valve #1    , via solid-state relay on pin D12
    - solenoid valve #2    , via solid-state relay on pin D5
    - external 220 VAC pump, via solid-state relay on pin D13

  The RGB LED of the Feather M4 will indicate its status:
    - Blue : We're setting up
    - Green: All okay and idling
  Every read out, the LED will flash brightly turquoise.

  The Arduino will autonomously read out the sensor data at fixed time intervals
  and will report these readings over serial.

  https://github.com/Dennis-van-Gils/project-Humidistat
  Dennis van Gils
  26-06-2023
*******************************************************************************/

#include <Arduino.h>
#include <SPI.h>
#include <Wire.h>

#include "Adafruit_BME280.h"
#include "Adafruit_NeoPixel.h"
#include "Adafruit_Sensor.h"
#include "DvG_SerialCommand.h"

// Pin assignments for actuators
#define PIN_VALVE_1 12
#define PIN_VALVE_2 5
#define PIN_PUMP 13

// Figure out which Adafruit board is being used: M0 or M4
#if defined _VARIANT_ARDUINO_ZERO_
#  define HAS_NEOPIXEL 0
#elif defined _VARIANT_FEATHER_M4_
#  define HAS_NEOPIXEL 1
#endif

// BME280: Temperature, humidity and pressure sensors
// NOTE: Do not read out faster than once per second as per BME280 spec-sheet.
#define DAQ_PERIOD 1000 // [ms] Data-acquisition period
Adafruit_BME280 bme_1;
Adafruit_BME280 bme_2;

#if HAS_NEOPIXEL == 1
// On-board NeoPixel RGB LED
#  define NEO_DIM 3    // Brightness level for dim intensity [0 -255]
#  define NEO_BRIGHT 6 // Brightness level for bright intensity [0 - 255]
#  define NEO_FLASH_DURATION 100 // [ms]
bool neo_flash = false;
uint32_t t_neo_flash = 0;
Adafruit_NeoPixel neo(1, PIN_NEOPIXEL, NEO_GRB + NEO_KHZ800);
#endif

// Serial command listener
DvG_SerialCommand sc(Serial);

// -----------------------------------------------------------------------------
//  State control
// -----------------------------------------------------------------------------

struct State {
  /* Holds actual actuator states and sensor readings
   */
  bool valve_1 = false;
  bool valve_2 = false;
  bool pump = false;
  float temp_1 = NAN; // ['C]
  float temp_2 = NAN; // ['C]
  float humi_1 = NAN; // [% RH]
  float humi_2 = NAN; // [% RH]
  float pres_1 = NAN; // [Pa]
  float pres_2 = NAN; // [Pa]
};
State state;

struct Request {
  /* Holds requested actuator states
   */
  bool valve_1 = false;
  bool valve_2 = false;
  bool pump = false;
};
Request request;

void connect_BME280_sensors() {
  uint8_t idx_try;
  bool success;

  for (idx_try = 0; idx_try < 3; idx_try++) {
    success = bme_1.begin(0x76);
    if (success) {
      break;
    }
    delay(1000);
  }
  if (!success) {
    Serial.println("Could not find BME280 sensor #1");
  }

  for (idx_try = 0; idx_try < 3; idx_try++) {
    success = bme_2.begin(0x77);
    if (success) {
      break;
    }
    delay(1000);
  }
  if (!success) {
    Serial.println("Could not find BME280 sensor #2");
  }
}

void read_BME280_sensors() {
  /* Update the sensor readings and save in `state`.
  NOTE: Do not read out faster than once per second as per BME280 spec-sheet.
   */
  state.temp_1 = bme_1.readTemperature();
  state.humi_1 = bme_1.readHumidity();
  state.pres_1 = bme_1.readPressure();

  state.temp_2 = bme_2.readTemperature();
  state.humi_2 = bme_2.readHumidity();
  state.pres_2 = bme_2.readPressure();
}

void report(uint32_t now, uint32_t t_0) {
  // Report `state` over serial
  Serial.print(now - t_0);
  Serial.write('\t');
  Serial.print(state.valve_1);
  Serial.write('\t');
  Serial.print(state.valve_2);
  Serial.write('\t');
  Serial.print(state.pump);
  Serial.write('\t');
  Serial.print(state.humi_1, 2);
  Serial.write('\t');
  Serial.print(state.humi_2, 2);
  Serial.write('\t');
  Serial.print(state.temp_1, 2);
  Serial.write('\t');
  Serial.print(state.temp_2, 2);
  Serial.write('\t');
  Serial.print(state.pres_1, 0);
  Serial.write('\t');
  Serial.print(state.pres_2, 0);
  Serial.write('\n');
}

void measure_and_report(uint32_t now, uint32_t t_0) {
  /* Perform a single measurement, update `state` and report over serial.
   */

#if HAS_NEOPIXEL == 1
  // Set RGB LED to bright turquoise: Performing new measurement
  neo_flash = true;
  t_neo_flash = now;
  neo.setPixelColor(0, neo.Color(0, NEO_BRIGHT, NEO_BRIGHT));
  neo.show();
#endif

  read_BME280_sensors();
  report(now, t_0);
}

bool parseBoolInString(char *strIn, uint8_t iPos) {
  if (strlen(strIn) > iPos) {
    return (strncmp(&strIn[iPos], "1", 1) == 0);
  } else {
    return false;
  }
}

int parseIntInString(char *strIn, uint8_t iPos) {
  if (strlen(strIn) > iPos) {
    return atoi(&strIn[iPos]);
  } else {
    return 0;
  }
}

// -----------------------------------------------------------------------------
//  setup
// -----------------------------------------------------------------------------

void setup() {
  // Initialize actuators
  pinMode(PIN_VALVE_1, OUTPUT);
  pinMode(PIN_VALVE_2, OUTPUT);
  pinMode(PIN_PUMP, OUTPUT);
  digitalWrite(PIN_VALVE_1, state.valve_1);
  digitalWrite(PIN_VALVE_2, state.valve_2);
  digitalWrite(PIN_PUMP, state.pump);

#if HAS_NEOPIXEL == 1
  // Set RGB LED to blue: We're setting up
  neo.begin();
  neo.setPixelColor(0, neo.Color(0, 0, NEO_BRIGHT));
  neo.show();
#endif

  Serial.begin(9600);

  connect_BME280_sensors();
  read_BME280_sensors(); // Ditch the first reading. Tends to be off.

#if HAS_NEOPIXEL == 1
  // Set RGB LED to dim green: We're all ready to go and idle
  neo.setPixelColor(0, neo.Color(0, NEO_DIM, 0));
  neo.show();
#endif
}

// -----------------------------------------------------------------------------
//  loop
// -----------------------------------------------------------------------------

void loop() {
  char *str_cmd; // Incoming serial command string
  uint32_t now = millis();
  static uint32_t t_0 = now;
  static uint32_t tick = now;

  // Short burst control
  static bool burst = false;   // Burst in progress?
  static uint32_t T_burst = 0; // [ms] Length of burst
  static uint32_t t_burst = 0; // `millis()` start of burst

  // Process incoming serial commands
  if (sc.available()) {
    str_cmd = sc.getCmd();

    if (strcmp(str_cmd, "id?") == 0) {
      Serial.println("Arduino, Humidistat v1");

    } else if (strncmp(str_cmd, "a", 1) == 0) {
      // Request new actuator states all at once
      // Command parameters:
      //  "a"[0/1: valve_1?][0/1: valve_2?][0/1: pump?]
      //   E.g: "a101" to open valve_1, close valve_2 and enable the pump
      request.valve_1 = parseBoolInString(str_cmd, 1);
      request.valve_2 = parseBoolInString(str_cmd, 2);
      request.pump = parseBoolInString(str_cmd, 3);

    } else if (strncmp(str_cmd, "b", 1) == 0) {
      // `Burst` mode: Request open valve 1 and/or valve 2 and/or the pump for a
      // short fixed time duration. Closes all again after the elapsed time.
      // Command parameters:
      //  "b"[0/1: valve_1?][0/1: valve_2?][0/1: pump?][int: duration in ms]
      //   E.g: "b101500" to open valve_1 and the pump for 500 ms
      request.valve_1 = parseBoolInString(str_cmd, 1);
      request.valve_2 = parseBoolInString(str_cmd, 2);
      request.pump = parseBoolInString(str_cmd, 3);
      T_burst = parseIntInString(str_cmd, 4);
      t_burst = now;
      burst = true;

    } else if (strncmp(str_cmd, "v1", 2) == 0) {
      // Turn valve 1 on/off
      request.valve_1 = parseBoolInString(str_cmd, 2);

    } else if (strncmp(str_cmd, "v2", 2) == 0) {
      // Turn valve 2 on/off
      request.valve_2 = parseBoolInString(str_cmd, 2);

    } else if (strncmp(str_cmd, "p", 1) == 0) {
      // Turn pump on/off
      request.pump = parseBoolInString(str_cmd, 1);

    } else if (strcmp(str_cmd, "r") == 0) {
      // Try to reconnect to BME280's
      connect_BME280_sensors();
    }
  }

  // Request to turn off actuators at end of burst
  if (burst && (now - t_burst >= T_burst)) {
    burst = false;
    request.valve_1 = false;
    request.valve_2 = false;
    request.pump = false;
  }

  // Grant all requested actuator states
  bool change_detected = false;
  if (request.valve_1 != state.valve_1) {
    state.valve_1 = request.valve_1;
    digitalWrite(PIN_VALVE_1, state.valve_1);
    change_detected = true;
  }
  if (request.valve_2 != state.valve_2) {
    state.valve_2 = request.valve_2;
    digitalWrite(PIN_VALVE_2, state.valve_2);
    change_detected = true;
  }
  if (request.pump != state.pump) {
    state.pump = request.pump;
    digitalWrite(PIN_PUMP, state.pump);
    change_detected = true;
  }

  // Report immediately over serial when actuators have changed state
  if (change_detected) {
    report(now, t_0);
  }

  // DAQ
  if (now - tick >= DAQ_PERIOD) {
    tick += DAQ_PERIOD; // Strict-interval time keeping
    measure_and_report(now, t_0);
  }

#if HAS_NEOPIXEL == 1
  // Set RGB LED back to dim green: Measurement is done
  if (neo_flash && (now - t_neo_flash >= NEO_FLASH_DURATION)) {
    neo_flash = false;
    neo.setPixelColor(0, neo.Color(0, NEO_DIM, 0));
    neo.show();
  }
#endif
}


src_python/humidistat_main.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""Humidistat

A humidity controller for fluid dynamics research
"""
__author__ = "Dennis van Gils"
__authoremail__ = "vangils.dennis@gmail.com"
__url__ = "https://github.com/Dennis-van-Gils/project-Humidistat"
__date__ = "22-01-2021"
__version__ = "1.0"
# pylint: disable=bare-except, broad-except

import sys
import time

from PyQt5 import QtCore
from PyQt5 import QtWidgets as QtWid
from PyQt5.QtCore import QDateTime
import numpy as np

from dvg_pyqt_filelogger import FileLogger
from dvg_debug_functions import dprint, print_fancy_traceback as pft

from dvg_devices.Arduino_protocol_serial import Arduino
from humidistat_qdev import Humidistat_qdev
from humidistat_gui import MainWindow


# Constants
DAQ_INTERVAL_MS = 1000  # [ms]

# Show debug info in terminal? Warning: Slow! Do not leave on unintentionally.
DEBUG = False


def current_date_time_strings():
    cur_date_time = QDateTime.currentDateTime()
    return (
        cur_date_time.toString("dd-MM-yyyy"),  # Date
        cur_date_time.toString("HH:mm:ss"),  # Time
    )


# ------------------------------------------------------------------------------
#   Program termination routines
# ------------------------------------------------------------------------------


def stop_running():
    app.processEvents()
    ard_qdev.quit()
    logger.close()


@QtCore.pyqtSlot()
def notify_connection_lost():
    stop_running()

    window.qlbl_title.setText("! ! !    LOST CONNECTION    ! ! !")
    str_cur_date, str_cur_time = current_date_time_strings()
    str_msg = "%s %s\nLost connection to Arduino." % (
        str_cur_date,
        str_cur_time,
    )
    print("\nCRITICAL ERROR @ %s" % str_msg)
    reply = QtWid.QMessageBox.warning(
        window, "CRITICAL ERROR", str_msg, QtWid.QMessageBox.Ok
    )

    if reply == QtWid.QMessageBox.Ok:
        pass  # Leave the GUI open for read-only inspection by the user


@QtCore.pyqtSlot()
def about_to_quit():
    print("\nAbout to quit")
    stop_running()
    ard.close()


# ------------------------------------------------------------------------------
#   Arduino data-acquisition update function
# ------------------------------------------------------------------------------


def DAQ_function():
    # Shorthands
    state: ard_qdev.State = ard_qdev.state
    config: ard_qdev.Config = ard_qdev.config

    # Date-time keeping
    str_cur_date, str_cur_time = current_date_time_strings()

    # Query the Arduino for its state
    success, tmp_state = ard.query_ascii_values("?", delimiter="\t")
    if not (success):
        dprint(
            "'%s' reports IOError @ %s %s"
            % (ard.name, str_cur_date, str_cur_time)
        )
        return False

    # Parse readings into separate state variables
    try:
        (
            state.time,
            state.valve_1,
            state.valve_2,
            state.pump,
            state.humi_1,
            state.humi_2,
            state.temp_1,
            state.temp_2,
            state.pres_1,
            state.pres_2,
        ) = tmp_state
        state.valve_1 = bool(state.valve_1)
        state.valve_2 = bool(state.valve_2)
        state.pump = bool(state.pump)
        state.time /= 1000  # Arduino time, [msec] to [s]
        state.pres_1 /= 100  # [Pa] to [mbar]
        state.pres_2 /= 100  # [Pa] to [mbar]
    except Exception as err:
        pft(err, 3)
        dprint(
            "'%s' reports IOError @ %s %s"
            % (ard.name, str_cur_date, str_cur_time)
        )
        return False

    # We will use PC time instead
    state.time = time.perf_counter()

    # Add readings to chart histories
    window.curve_humi_1.appendData(state.time, state.humi_1)
    window.curve_temp_1.appendData(state.time, state.temp_1)
    window.curve_pres_1.appendData(state.time, state.pres_1)
    window.curve_humi_2.appendData(state.time, state.humi_2)
    window.curve_temp_2.appendData(state.time, state.temp_2)
    window.curve_pres_2.appendData(state.time, state.pres_2)

    # Logging to file
    logger.update(mode="w")

    # Return success
    return True


# ------------------------------------------------------------------------------
#   File logger
# ------------------------------------------------------------------------------


def write_header_to_log():
    logger.write("[HEADER]\n")
    logger.write(window.qtxt_comments.toPlainText())
    logger.write("\n\n[DATA]\n")
    logger.write(
        "[s]\t"
        "[0/1]\t[0/1]\t[0/1]\t"
        "[±3 pct]\t[±0.5 °C]\t[±1 mbar]\t"
        "[±3 pct]\t[±0.5 °C]\t[±1 mbar]\n"
    )
    logger.write(
        "time\t"
        "valve_1\tvalve_2\tpump\t"
        "humi_1\ttemp_1\tpres_1\t"
        "humi_2\ttemp_2\tpres_2\n"
    )


def write_data_to_log():
    state = ard_qdev.state  # Shorthand
    logger.write(
        "%.0f\t%u\t%u\t%u\t%.1f\t%.1f\t%.1f\t%.1f\t%.1f\t%.1f\n"
        % (
            logger.elapsed(),
            state.valve_1,
            state.valve_2,
            state.pump,
            state.humi_1,
            state.temp_1,
            state.pres_1,
            state.humi_2,
            state.temp_2,
            state.pres_2,
        )
    )


# ------------------------------------------------------------------------------
#   Main
# ------------------------------------------------------------------------------

if __name__ == "__main__":

    # Connect to Arduino
    ard = Arduino(name="Ard", connect_to_specific_ID="Humidistat v1")
    ard.serial_settings["baudrate"] = 115200
    ard.auto_connect(filepath_last_known_port="config/port_Arduino.txt")

    if not (ard.is_alive):
        print("\nCheck connection and try resetting the Arduino.")
        print("Exiting...\n")
        sys.exit(0)

    # Set up multi-threaded communication: Creates workers and threads
    ard_qdev = Humidistat_qdev(
        dev=ard,
        DAQ_function=DAQ_function,
        DAQ_interval_ms=DAQ_INTERVAL_MS,
        debug=DEBUG,
    )
    ard_qdev.signal_connection_lost.connect(notify_connection_lost)

    # File logger
    logger = FileLogger(
        write_header_function=write_header_to_log,
        write_data_function=write_data_to_log,
    )

    # Create application and main window
    QtCore.QThread.currentThread().setObjectName("MAIN")  # For DEBUG info
    app = QtWid.QApplication(sys.argv)
    app.aboutToQuit.connect(about_to_quit)
    window = MainWindow(ard, ard_qdev, logger)

    # Start threads
    ard_qdev.start()

    # Start the main GUI event loop
    window.show()
    sys.exit(app.exec_())


src_python/humidistat_qdev.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""humidistat_qdev.py

Manages multi-threaded communication with the Arduino
"""
__author__ = "Dennis van Gils"
__authoremail__ = "vangils.dennis@gmail.com"
__url__ = "https://github.com/Dennis-van-Gils/project-Humidistat"
__date__ = "28-07-2022"
__version__ = "1.1"

from enum import Enum
import numpy as np

from dvg_devices.Arduino_protocol_serial import Arduino
from dvg_qdeviceio import QDeviceIO, DAQ_TRIGGER


class ControlMode(Enum):
    Manual = 0
    Auto = 1


class ControlBand(Enum):
    Coarse = 0
    Fine = 1
    Dead = 2


class Actuators:
    """Holds which actuators to enable to either increase or decrease the
    humidity. Hence, two instances of this class should be created."""

    def __init__(
        self, valve_1: bool = False, valve_2: bool = False, pump: bool = False
    ):
        self.ENA_valve_1 = valve_1
        self.ENA_valve_2 = valve_2
        self.ENA_pump = pump


class Humidistat_qdev(QDeviceIO):
    class State(object):
        def __init__(self):
            # Actual readings of the Arduino
            self.time = np.nan  # [s]
            self.valve_1 = False
            self.valve_2 = False
            self.pump = False
            self.temp_1 = np.nan  # ['C]
            self.temp_2 = np.nan  # ['C]
            self.humi_1 = np.nan  # [% RH]
            self.humi_2 = np.nan  # [% RH]
            self.pres_1 = np.nan  # [mbar]
            self.pres_2 = np.nan  # [mbar]

            # Control
            self.setpoint = 50  # [% RH]
            self.control_mode = ControlMode.Manual
            self.control_band = ControlBand.Coarse
            self.control_band_prev = None
            self.t_burst = 0  # [s], timestamp at start of burst period

    class Config(object):
        def __init__(self):
            # fmt: off
            # Actuators
            self.actors_incr_RH = Actuators(True, False, True)
            self.actors_decr_RH = Actuators(False, True, True)
            self.act_on_sensor_no = 1  # [1 or 2]

            # Bandwidths
            self.fineband_dHI = +2     # [% RH]
            self.fineband_dLO = -2     # [% RH]
            self.deadband_dHI = +0.5   # [% RH]
            self.deadband_dLO = -0.5   # [% RH]

            # Fine 'burst' control mode
            self.burst_update_period = 10     # [s]
            self.burst_incr_RH_length = 1000  # [ms]
            self.burst_decr_RH_length = 1000  # [ms]
            # fmt: on

    # --------------------------------------------------------------------------
    #   Humidistat_qdev
    # --------------------------------------------------------------------------

    def __init__(
        self,
        dev: Arduino,
        DAQ_function=None,
        debug=False,
        **kwargs,
    ):
        super().__init__(dev, **kwargs)  # Pass kwargs onto QtCore.QObject()

        self.state = self.State()
        self.config = self.Config()

        self.create_worker_DAQ(
            DAQ_trigger=DAQ_TRIGGER.CONTINUOUS,
            DAQ_function=DAQ_function,
            critical_not_alive_count=3,
            debug=debug,
        )
        self.create_worker_jobs(debug=debug)

    # --------------------------------------------------------------------------
    #   Arduino communication functions
    # --------------------------------------------------------------------------

    def reconnect_BME280_sensors(self):
        self.send(self.dev.write, "r")

    def set_valve_1(self, flag: bool):
        if not self.state.valve_1 == flag:
            self.send(self.dev.write, "v1%u" % flag)

    def set_valve_2(self, flag: bool):
        if not self.state.valve_2 == flag:
            self.send(self.dev.write, "v2%u" % flag)

    def set_pump(self, flag: bool):
        if not self.state.pump == flag:
            self.send(self.dev.write, "p%u" % flag)

    def set_actuators(self, valve_1: bool, valve_2: bool, pump: bool):
        if (
            (not self.state.valve_1 == valve_1)
            or (not self.state.valve_2 == valve_2)
            or (not self.state.pump == pump)
        ):
            self.send(self.dev.write, "a%u%u%u" % (valve_1, valve_2, pump))

    def burst_incr_RH(self):
        command = "b%u%u%u%u" % (
            self.config.actors_incr_RH.ENA_valve_1,
            self.config.actors_incr_RH.ENA_valve_2,
            self.config.actors_incr_RH.ENA_pump,
            self.config.burst_incr_RH_length,
        )
        self.send(self.dev.write, command)

    def burst_decr_RH(self):
        command = "b%u%u%u%u" % (
            self.config.actors_decr_RH.ENA_valve_1,
            self.config.actors_decr_RH.ENA_valve_2,
            self.config.actors_decr_RH.ENA_pump,
            self.config.burst_decr_RH_length,
        )
        self.send(self.dev.write, command)


src_python/main.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""Humidistat

A humidity controller for fluid dynamics research
"""
__author__ = "Dennis van Gils"
__authoremail__ = "vangils.dennis@gmail.com"
__url__ = "https://github.com/Dennis-van-Gils/project-Humidistat"
__date__ = "28-07-2022"
__version__ = "1.1"
# pylint: disable=bare-except, broad-except

import sys
import time

from PyQt5 import QtCore
from PyQt5 import QtWidgets as QtWid
from PyQt5.QtCore import QDateTime

from dvg_pyqt_filelogger import FileLogger
from dvg_debug_functions import dprint, print_fancy_traceback as pft

from dvg_devices.Arduino_protocol_serial import Arduino
from humidistat_qdev import Humidistat_qdev, ControlMode, ControlBand
from humidistat_gui import MainWindow

# Show debug info in terminal?
DEBUG = False

# ------------------------------------------------------------------------------
#   current_date_time_strings
# ------------------------------------------------------------------------------


def current_date_time_strings():
    cur_date_time = QDateTime.currentDateTime()
    return (
        cur_date_time.toString("dd-MM-yyyy"),  # Date
        cur_date_time.toString("HH:mm:ss"),  # Time
    )


# ------------------------------------------------------------------------------
#   Program termination routines
# ------------------------------------------------------------------------------


def stop_running():
    app.processEvents()
    ard_qdev.quit()
    logger.close()


@QtCore.pyqtSlot()
def notify_connection_lost():
    stop_running()

    window.qlbl_title.setText("! ! !    LOST CONNECTION    ! ! !")
    str_cur_date, str_cur_time = current_date_time_strings()
    str_msg = "%s %s\nLost connection to Arduino." % (
        str_cur_date,
        str_cur_time,
    )
    print("\nCRITICAL ERROR @ %s" % str_msg)
    reply = QtWid.QMessageBox.warning(
        window, "CRITICAL ERROR", str_msg, QtWid.QMessageBox.Ok
    )

    if reply == QtWid.QMessageBox.Ok:
        pass  # Leave the GUI open for read-only inspection by the user


@QtCore.pyqtSlot()
def about_to_quit():
    print("\nAbout to quit")
    stop_running()
    ard.close()


# ------------------------------------------------------------------------------
#   Arduino data-acquisition update function
# ------------------------------------------------------------------------------


def DAQ_function() -> bool:
    # WARNING: Do not change the GUI directly from out of this function as it
    # will be running in a separate and different thread to the main/GUI thread.

    # Shorthands
    state = ard_qdev.state
    config = ard_qdev.config

    # Date-time keeping
    str_cur_date, str_cur_time = current_date_time_strings()

    # Listen to the Arduino for sensor readings send out over serial
    success, reply = ard.readline()
    dprint(reply)

    if not (success):
        dprint(
            "'%s' reports IOError @ %s %s"
            % (ard.name, str_cur_date, str_cur_time)
        )
        return False

    # Parse readings into separate state variables
    try:
        (
            state.time,
            state.valve_1,
            state.valve_2,
            state.pump,
            state.humi_1,
            state.humi_2,
            state.temp_1,
            state.temp_2,
            state.pres_1,
            state.pres_2,
        ) = list(
            map(float, reply.split("\t"))
        )  # Parse all as float to allow for "nan" values in the reply string
        state.valve_1 = bool(state.valve_1)
        state.valve_2 = bool(state.valve_2)
        state.pump = bool(state.pump)
        state.time /= 1000  # Arduino time, [msec] to [s]
        state.pres_1 /= 100  # [Pa] to [mbar]
        state.pres_2 /= 100  # [Pa] to [mbar]
    except Exception as err:
        pft(err)
        dprint(
            "'%s' reports IOError @ %s %s"
            % (ard.name, str_cur_date, str_cur_time)
        )
        return False

    # We will use PC time instead
    state.time = time.perf_counter()

    # Measurement bandwidth control mechanism
    humi = state.humi_1 if config.act_on_sensor_no == 1 else state.humi_2
    humi_err = humi - state.setpoint

    if (humi_err > config.deadband_dLO) & (humi_err < config.deadband_dHI):
        state.control_band = ControlBand.Dead
        if state.control_band != state.control_band_prev:
            dprint("Control band: DEAD")
    elif (humi_err > config.fineband_dLO) & (humi_err < config.fineband_dHI):
        state.control_band = ControlBand.Fine
        if state.control_band != state.control_band_prev:
            dprint("Control band: FINE")
    else:
        state.control_band = ControlBand.Coarse
        if state.control_band != state.control_band_prev:
            dprint("Control band: COARSE")

    if state.control_mode == ControlMode.Auto:
        if state.control_band == ControlBand.Coarse:
            if humi < state.setpoint:
                ard_qdev.set_actuators(
                    config.actors_incr_RH.ENA_valve_1,
                    config.actors_incr_RH.ENA_valve_2,
                    config.actors_incr_RH.ENA_pump,
                )
            else:
                ard_qdev.set_actuators(
                    config.actors_decr_RH.ENA_valve_1,
                    config.actors_decr_RH.ENA_valve_2,
                    config.actors_decr_RH.ENA_pump,
                )

        elif state.control_band == ControlBand.Fine:
            if state.control_band != state.control_band_prev:
                # Restart burst timer as soon as we enter the fine-band and
                # ensure we turn off all actuators
                state.t_burst = time.perf_counter()
                ard_qdev.set_actuators(False, False, False)

            if time.perf_counter() - state.t_burst > config.burst_update_period:
                # Burst timer fired
                if humi < state.setpoint:
                    ard_qdev.burst_incr_RH()
                else:
                    ard_qdev.burst_decr_RH()

                state.t_burst = time.perf_counter()

        else:
            # Dead-band: Turn off all actuators
            ard_qdev.set_actuators(False, False, False)

    state.control_band_prev = state.control_band

    # Add readings to chart histories
    window.curve_humi_1.appendData(state.time, state.humi_1)
    window.curve_temp_1.appendData(state.time, state.temp_1)
    window.curve_pres_1.appendData(state.time, state.pres_1)
    window.curve_humi_2.appendData(state.time, state.humi_2)
    window.curve_temp_2.appendData(state.time, state.temp_2)
    window.curve_pres_2.appendData(state.time, state.pres_2)
    window.curve_setpoint.appendData(state.time, state.setpoint)
    window.curve_deadband_HI.appendData(
        state.time, state.setpoint + config.deadband_dHI
    )
    window.curve_deadband_LO.appendData(
        state.time, state.setpoint + config.deadband_dLO
    )

    # Logging to file
    logger.update(mode="w")

    # Return success
    return True


# ------------------------------------------------------------------------------
#   File logger
# ------------------------------------------------------------------------------


def write_header_to_log():
    str_cur_date, str_cur_time = current_date_time_strings()
    logger.write("[HEADER]\n")
    logger.write(str_cur_date + "\n")
    logger.write(str_cur_time + "\n")
    logger.write(window.qtxt_comments.toPlainText())
    logger.write("\n\n[DATA]\n")
    logger.write(
        "[s]\t"
        "[0/1]\t[0/1]\t[0/1]\t"
        "[±3 pct]\t[±0.5 °C]\t[±1 mbar]\t"
        "[±3 pct]\t[±0.5 °C]\t[±1 mbar]\n"
    )
    logger.write(
        "time\t"
        "valve_1\tvalve_2\tpump\t"
        "humi_1\ttemp_1\tpres_1\t"
        "humi_2\ttemp_2\tpres_2\n"
    )


def write_data_to_log():
    state = ard_qdev.state  # Shorthand
    logger.write(
        "%.3f\t%u\t%u\t%u\t%.2f\t%.2f\t%.1f\t%.2f\t%.2f\t%.1f\n"
        % (
            logger.elapsed(),
            state.valve_1,
            state.valve_2,
            state.pump,
            state.humi_1,
            state.temp_1,
            state.pres_1,
            state.humi_2,
            state.temp_2,
            state.pres_2,
        )
    )


# ------------------------------------------------------------------------------
#   Main
# ------------------------------------------------------------------------------

if __name__ == "__main__":

    # Connect to Arduino
    ard = Arduino(name="Ard", connect_to_specific_ID="Humidistat v1")
    ard.serial_settings["baudrate"] = 115200
    ard.auto_connect(filepath_last_known_port="config/port_Arduino.txt")

    if not (ard.is_alive):
        print("\nCheck connection and try resetting the Arduino.\n")
        # print("Exiting...\n")
        # sys.exit(0)

    # Set up multi-threaded communication: Creates workers and threads
    ard_qdev = Humidistat_qdev(
        dev=ard,
        DAQ_function=DAQ_function,
        debug=DEBUG,
    )
    ard_qdev.signal_connection_lost.connect(notify_connection_lost)

    # File logger
    logger = FileLogger(
        write_header_function=write_header_to_log,
        write_data_function=write_data_to_log,
    )

    # Create application and main window
    QtCore.QThread.currentThread().setObjectName("MAIN")  # For DEBUG info
    app = QtWid.QApplication(sys.argv)
    app.aboutToQuit.connect(about_to_quit)
    window = MainWindow(ard, ard_qdev, logger)

    # Start threads
    ard_qdev.start()  # NOTE: The DAQ worker thread starts up in a paused state
    ard_qdev.unpause_DAQ()  # Start listening continuously to the Arduino

    # Load default Humidistat configuration from file
    window.load_config_from_file(from_default=True)

    # Start the main GUI event loop
    window.show()
    sys.exit(app.exec_())
